### PR TITLE
[wasm] Extract `src/mono/browser` from `src/mono/wasm`

### PR DESCRIPTION
**Pull request by radical**
- `Wasm.Build.Tests` is still in `src/mono/wasm`, mainly because it has lot of common code for build tests. And this will share more code with wasi-build-tests in future PRs.
- This PR adds a `src/mono/browser`, and then moves some browser specific directories from `src/mono/wasm` to it.
- Then the corresponding references to the paths are updated in all the places in the repo.

Includes changes based on review feedback from @ maraf, and @ pavelsavara

### CODE REVIEW

File: src/mono/browser/wasm.proj
Comment (maraf) [inline src/mono/browser/wasm.proj:1]: Is it the right time to rename the project file?

File: src/mono/mono.proj [hunk 1]
```diff
@@ -149,15 +149,15 @@
   <!-- Sets up emscripten if you don't have the EMSDK_PATH env variable set -->
   <Target Name="ProvisionEmscripten"
```
Comment (maraf) [inline src/mono/mono.proj:150]: Unrelated: why is emscripten provisioning in mono.proj instead of wasm.proj?

File: src/mono/wasi/wasi.proj [hunk 2]
```diff
@@ -226,17 +226,17 @@
...
       <_FilesToCopy Include="$(MSBuildThisFileDirectory)runtime/stubs.c"             DestinationFolder="$(NativeBinDir)src" />
       <_FilesToCopy Include="$(MSBuildThisFileDirectory)runtime/synthetic-pthread.c" DestinationFolder="$(NativeBinDir)src" />
 
-      <_FilesToCopy Include="$(MonoProjectRoot)wasi\mono-include\driver.h"           DestinationFolder="$(NativeBinDir)include\wasm" />
-      <_FilesToCopy Include="$(MonoProjectRoot)wasm\runtime\gc-common.h"             DestinationFolder="$(NativeBinDir)include\wasm" />
-      <_FilesToCopy Include="$(MSBuildThisFileDirectory)..\wasm\runtime\runtime.h"   DestinationFolder="$(NativeBinDir)include\wasm" />
-      <_FilesToCopy Include="$(MSBuildThisFileDirectory)..\wasm\runtime\pinvoke.h"   DestinationFolder="$(NativeBinDir)include\wasm" />
+      <_FilesToCopy Include="$(WasiProjectRoot)mono-include\driver.h"           DestinationFolder="$(NativeBinDir)include\wasm" />
+      <_FilesToCopy Include="$(BrowserProjectRoot)runtime\gc-common.h"               DestinationFolder="$(NativeBinDir)include\wasm" />
```
Comment (maraf) [inline src/mono/wasi/wasi.proj:237]: Should we move these sources to common directory?

File: src/mono/wasm/build/WasmApp.LocalBuild.props [hunk 3]
```diff
@@ -25,12 +25,11 @@
     <_TargetFrameworkForNETCoreTasks>net8.0</_TargetFrameworkForNETCoreTasks>
     <MSBuildEnableWorkloadResolver>false</MSBuildEnableWorkloadResolver>
     <SelfContained>true</SelfContained>
-    <WasmCommonTargetsPath>$(_WasmTargetsDir)</WasmCommonTargetsPath>
   </PropertyGroup>
 
   <PropertyGroup Condition="'$(RuntimeSrcDir)' != '' and '$(WasmBuildSupportDir)' == ''">
     <ArtifactsBinDir>$(RuntimeSrcDir)\artifacts\bin\</ArtifactsBinDir>
-    <EMSDK_PATH Condition="'$(EMSDK_PATH)' == ''">$(RuntimeSrcDir)\src\mono\wasm\emsdk\</EMSDK_PATH>
+    <EMSDK_PATH Condition="'$(EMSDK_PATH)' == ''">$(RuntimeSrcDir)\src\mono\browser\emsdk\</EMSDK_PATH>
```
Comment (maraf) [inline src/mono/wasm/build/WasmApp.LocalBuild.props:32]: Can we use `BrowserProjectRoot` here??

```suggestion
    <EMSDK_PATH Condition="'$(EMSDK_PATH)' == ''">$(BrowserProjectRoot)\emsdk\</EMSDK_PATH>
```

File: src/mono/wasm/data/aot-tests/Directory.Build.props [hunk 4]
```diff
@@ -13,16 +13,20 @@
     <WASI_SDK_PATH>$(WasmBuildSupportDir)\wasi-sdk\</WASI_SDK_PATH>
   </PropertyGroup>
 
-  <PropertyGroup >
-    <_WasmTargetsDir Condition="'$(_WasmTargetsDir)' == '' and '$(RuntimeSrcDir)' != '' and '$(RuntimeIdentifier)' == 'browser-wasm'">$(RuntimeSrcDir)\src\mono\wasm\build\</_WasmTargetsDir>
+  <PropertyGroup Condition="'$(RuntimeSrcDir)' != ''">
+    <_WasmTargetsDir Condition="'$(_WasmTargetsDir)' == '' and '$(RuntimeSrcDir)' != '' and '$(RuntimeIdentifier)' == 'browser-wasm'">$(RuntimeSrcDir)\src\mono\browser\build\</_WasmTargetsDir>
```
Comment (maraf) [inline src/mono/wasm/data/aot-tests/Directory.Build.props:17]: Can we use `BrowserProjectRoot` here?

```suggestion
    <_WasmTargetsDir Condition="'$(_WasmTargetsDir)' == '' and '$(RuntimeSrcDir)' != '' and '$(RuntimeIdentifier)' == 'browser-wasm'">$(BrowserProjectRoot)\build\</_WasmTargetsDir>
```

File: src/tests/FunctionalTests/WebAssembly/Directory.Build.props [hunk 5]
```diff
@@ -28,7 +28,7 @@
   <PropertyGroup>
     <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
     <OutputPath>$(RepoRoot)artifacts\bin\$(MSBuildProjectName)</OutputPath>
-    <_InTreeEmSdk>$(MSBuildThisFileDirectory)..\..\..\mono\wasm\emsdk</_InTreeEmSdk>
+    <_InTreeEmSdk>$(MSBuildThisFileDirectory)..\..\..\mono\browser\emsdk</_InTreeEmSdk>
```
Comment (maraf) [inline src/tests/FunctionalTests/WebAssembly/Directory.Build.props:31]: Can we use `BrowserProjectRoot` here?

File: src/libraries/tests.proj [hunk 6]
```diff
@@ -683,7 +683,7 @@
                                      ('$(ContinuousIntegrationBuild)' != 'true' and '$(TestAssemblies)' == 'true'))"
                       BuildInParallel="false" />
 
-    <ProjectReference Include="$(RepoRoot)\src\mono\wasm\debugger\Wasm.Debugger.Tests\*.Tests.csproj"
+    <ProjectReference Include="$(RepoRoot)\src\mono\browser\debugger\Wasm.Debugger.Tests\*.Tests.csproj"
```
Comment (maraf) [inline src/libraries/tests.proj:686]: Can we use `BrowserProjectRoot` here?

File: src/libraries/sendtohelix-browser.targets [hunk 7]
```diff
@@ -51,7 +49,7 @@
     <EnableDefaultBuildHelixWorkItems>false</EnableDefaultBuildHelixWorkItems>
 
     <!-- on unix CI has emscripten provisioned in $(EMSDK_PATH) as `/usr/local/emscripten`. -->
-    <EMSDK_PATH Condition="$([MSBuild]::IsOSPlatform('WINDOWS')) and '$(EMSDK_PATH)' == ''">$(RepoRoot)src\mono\wasm\emsdk\</EMSDK_PATH>
+    <EMSDK_PATH Condition="$([MSBuild]::IsOSPlatform('WINDOWS')) and '$(EMSDK_PATH)' == ''">$(RepoRoot)src\mono\browser\emsdk\</EMSDK_PATH>
```
Comment (maraf) [inline src/libraries/sendtohelix-browser.targets:52]: Can we use `ProvisionEmscriptenDir` here?

File: src/mono/wasi/wasi.proj (hunk 2 above)
Comment (radical) [reply src/mono/wasi/wasi.proj:237]: Yeah, in follow up PRs we can do the necessary moving around.

File: src/mono/wasm/build/WasmApp.LocalBuild.props (hunk 3 above)
Comment (radical) [reply src/mono/wasm/build/WasmApp.LocalBuild.props:32]: This is *Localbuild.props which is meant to be used outside a repo where you wouldn't have it's `Directory.Build.props` available, thus no `BrowserProjectRoot`.

File: src/libraries/sendtohelix-browser.targets (hunk 7 above)
Comment (radical) [reply src/libraries/sendtohelix-browser.targets:52]: Good idea. We can move that property to the top level `Directory.Build.props`, and then use it here.

File: src/mono/mono.proj (hunk 1 above)
Comment (radical) [reply src/mono/mono.proj:150]: @radekdoulik would know better. My guess is that it just happens to be there, and we can move it to wasm.proj .

File: src/mono/browser/wasm.proj
Comment (radical) [reply src/mono/browser/wasm.proj:1]: Fixed.

File: src/mono/wasm/data/aot-tests/Directory.Build.props (hunk 4 above)
Comment (radical) [reply src/mono/wasm/data/aot-tests/Directory.Build.props:17]: This gets used outside the repo, on helix to be specific, so the property won't be set there.

**Comment by radical** [comment]
Perf run for sanity check - https://dev.azure.com/dnceng/internal/_build/results?buildId=2338052

File: src/mono/browser/.editorconfig [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+# top-most EditorConfig file
+root = false
+
+# Default settings:
+# A newline ending every file
+# Use 4 spaces as indentation
+[*]
+insert_final_newline = true
+indent_style = tabs
```
Comment (pavelsavara) [inline src/mono/browser/.editorconfig:11]: we don't use tabs in TS, do we really want to merge this file ?

Comment (radical) [reply src/mono/browser/.editorconfig:11]: oh, that looks wrong! good catch!

File: src/libraries/sendtohelix-browser.targets (hunk 7 above)
Comment (radical) [reply src/libraries/sendtohelix-browser.targets:52]: Though the property put in D.B.props should be `$(InTreeEmSdk)`, and that can be used in src/tests/FunctionalTests/WebAssembly/Directory.Build.props too.

**Review by maraf — APPROVED** [review]
Looks good to me 👍

**Pull request merged**